[DDP] Refactor hook running logic

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61020
* #61019
* **#61018**
* #61017

Extract logic of hook running to a function `run_reduction_hook` that takes in a `GradBucket` and runs the hook/allreduce. This is mainly to prepare for join to support comm. hook in follow up diffs.

Differential Revision: [D29477143](https://our.internmc.facebook.com/intern/diff/D29477143/)

rohan-varma committed:
[DDP] Refactor hook running logic

Extract logic of hook running to a function `run_reduction_hook` that takes in a `GradBucket` and runs the hook/allreduce. This is mainly to prepare for join to support comm. hook in follow up diffs.

Differential Revision: [D29477143](https://our.internmc.facebook.com/intern/diff/D29477143/)

[ghstack-poisoned]

SciPioneer commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -839,6 +839,18 @@ void Reducer::mark_variable_ready(size_t variable_index) {
 }
 
+c10::intrusive_ptr<c10::ivalue::Future> Reducer::run_reduction_hook(
+    GradBucket grad_bucket) {
```
Nit: pass it by reference, since it has a few vectors.

SciPioneer commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -105,6 +105,10 @@ class TORCH_API Reducer {
   void register_builtin_comm_hook(c10d::BuiltinCommHookType comm_hook_type);
 
+  // Runs allreduce or installed communication hook given GradBucket instance.
+  c10::intrusive_ptr<c10::ivalue::Future> run_reduction_hook(
```
Shall we define it as a private method instead? I expect this is just an internal helper function right?

rohan-varma replied:
I am planning to call this in `distributed.py` so it at least needs to be pybinded as a public method.

SciPioneer replied:
Does this mean that you want to be able to kick off a separate allreduce at somewhere else?

rohan-varma replied:
Yeah, when running with uneven inputs context manager, some of the processes in the entire world can already be finished training. To avoid hangs we then simulate an allreduce with zero tensors to match the allreduce of the currently active trainers.

SciPioneer commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -105,6 +105,10 @@ class TORCH_API Reducer {
   void register_builtin_comm_hook(c10d::BuiltinCommHookType comm_hook_type);
 
+  // Runs allreduce or installed communication hook given GradBucket instance.
+  c10::intrusive_ptr<c10::ivalue::Future> run_reduction_hook(
```
I think `run_comm_hook` is a better name. Please don't introduce a new concept of reduction hook.

rohan-varma committed:
Update on "[DDP] Refactor hook running logic"


Extract logic of hook running to a function `run_reduction_hook` that takes in a `GradBucket` and runs the hook/allreduce. This is mainly to prepare for join to support comm. hook in follow up diffs.

Differential Revision: [D29477143](https://our.internmc.facebook.com/intern/diff/D29477143/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor hook running logic"


Extract logic of hook running to a function `run_reduction_hook` that takes in a `GradBucket` and runs the hook/allreduce. This is mainly to prepare for join to support comm. hook in follow up diffs.

Differential Revision: [D29477143](https://our.internmc.facebook.com/intern/diff/D29477143/)

[ghstack-poisoned]

rohan-varma commented:
ROCM ci is completed, but looks like it is not reflected on the PR
